treewide: more accurate pandoc availability check

alyssais opened the pull request:
The actual problem here is that there is no GHC bootstrap tarball for RISC-V or LoongArch, so the right thing to check here is whether the platform being used to build GHC has a bootstrap tarball available for it.  This way, we'll do the right thing in all cases where such a tarball isn't available, not just riscv64 and loongarch64.

Since we need to refer to (unspliced) buildPackages.pandoc in the enableManpages default, I've opted to remove the pandoc input that would be spliced in nativeBuildInputs and use buildPackages.pandoc explicitly there as well, to avoid confusingly having two different pandocs around.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alyssais committed:
treewide: more accurate pandoc/shellcheck availability checks

The actual problem here is that there is no GHC bootstrap tarball for
RISC-V or LoongArch, so the right thing to check here is whether the
platform being used to build GHC has a bootstrap tarball available for
it.  This way, we'll do the right thing in all cases where such a
tarball isn't available, not just riscv64 and loongarch64, without
having to resort to tryEval, which could be hiding all sorts of
problems.

Since we need to refer to (unspliced) buildPackages.pandoc and
buildPackages.shellcheck in the conditionals, I've opted to remove the
pandoc and shellcheck inputs that would be spliced in
nativeBuildInputs and use buildPackages explicitly there as well, to
avoid confusingly having two different instances of the same package
around.

sternenseemann commented:
This is indeed the correct way to do it, though I wonder whether we should expose this as a purpose built function from `ghc.passthru` so this snippet doesn't need to be duplicated everywhere?

alyssais commented:
Putting it is passthru makes me nervous it wouldn't do the right thing in obscure cross situations, much like we have meta.available but know that it's broken and don't use it.

alyssais committed:
ghc.bootstrapAvailable: init

This will give packages that e.g. want to optionally disable pandoc
based on availability a simple attribute to check, rather than having
to copy quite a complex conditional between packages.

alyssais commented:
I've convinced myself that an attribute on GHC is actually okay.  The problem with pkg.meta.available is that you want to know whether something is available in the context of your current stdenv, but it doesn't know which stdenv you care about because it's just an attribute on another package.  In this case however, what we care about is whether the bootstrap is available in the context of GHC's stdenv, so having it as an attribute on GHC (as opposed to directly on the bootstrap) is fine.

misuzu commented:
See also
https://github.com/NixOS/nixpkgs/blob/190d40df762ab133c455fe73885eb749f510bca3/pkgs/build-support/trivial-builders/default.nix#L373
https://github.com/NixOS/nixpkgs/blob/190d40df762ab133c455fe73885eb749f510bca3/pkgs/by-name/ni/nixos-firewall-tool/package.nix#L30

sternenseemann commented:
Can you add the attribute to `common-make-native-bignum.nix` and `8.10.7.nix` as well, for consistency?

alyssais commented:
Done.

alyssais commented:
> See also
> 
> https://github.com/NixOS/nixpkgs/blob/190d40df762ab133c455fe73885eb749f510bca3/pkgs/build-support/trivial-builders/default.nix#L373
> 
> https://github.com/NixOS/nixpkgs/blob/190d40df762ab133c455fe73885eb749f510bca3/pkgs/by-name/ni/nixos-firewall-tool/package.nix#L30

Changed these too.
